lib/SILGen/SILGenExpr.cpp

```diff
@@ -2351,9 +2350,11 @@ static bool mayLieAboutNonOptionalReturn(Expr *expr) {
     return mayLieAboutNonOptionalReturn(declRef->getDecl());
   }
 
-  // An application, which we look through to get the function we're calling.
+  // An application that produces a reference type, which we look through to
+  // get the function we're calling.
   if (auto apply = dyn_cast<ApplyExpr>(expr)) {
-    return mayLieAboutNonOptionalReturn(apply->getFn());
+    return apply->getType()->hasReferenceSemantics() &&
```

Comment:
Yeah, that's what I meant: it looks like we already get this wrong. I'm not happy about us now checking in *two* places; it seems redundant.